radio_beam/commonbeam.py

```diff
@@ -0,0 +1,522 @@
+
+import numpy as np
+import astropy.units as u
+
+try:
+    from scipy import optimize as opt
+    from scipy.spatial import ConvexHull
+    HAS_SCIPY = True
+except ImportError:
+    HAS_SCIPY = False
+
+from .beam import Beam
+from .utils import BeamError, transform_ellipse
+
+
+def commonbeam(beams, method='pts', **method_kwargs):
+    '''
+    Use analytic method if there are only two beams. Otherwise use constrained
+    optimization to find the common beam.
+    '''
+
+    if beams.size == 1:
+        return beams[0]
+    elif fits_in_largest(beams):
+        return beams.largest_beam()
+    else:
+        if beams.size == 2:
+            try:
+                return common_2beams(beams)
+            # Sometimes this method can fail. Use the many beam solution in
+            # this case
+            except (ValueError, BeamError):
+                pass
+
+        if method == 'pts':
+            return common_manybeams_mve(beams, **method_kwargs)
+        elif method == 'opt':
+            return common_manybeams_opt(beams, **method_kwargs)
+        else:
+            raise ValueError("method must be 'pts' or 'opt'.")
+
+
+def common_2beams(beams):
+    '''
+    Find a common beam from a `Beams` object with 2 beams. This
+    function is based on the CASA implementation `ia.commonbeam`. Note that
+    the solution is only valid for 2 beams.
```

Comment:
Is there any case in which `reduce(common_2beams, beams)` would not result in the correct minimum enclosing beam?

Reply:
Yeah, the order that the beams are compared can change the size of the common beam. Using the CASA common beam function:
```
im1 = ia.newimagefromarray(pixels=ia.makearray(0, [4, 4, 1, 4]))
for i, pa in enumerate([0, 20, 40, 60]):
    im1.setrestoringbeam(channel=i, major='4arcsec', minor='3arcsec',
                         pa="{}deg".format(pa))

im2 = ia.newimagefromarray(pixels=ia.makearray(0, [4, 4, 1, 4]))
for i, pa in enumerate([0, 60, 20, 40]):
    im2.setrestoringbeam(channel=i, major='4arcsec', minor='3arcsec',
                         pa="{}deg".format(pa))

print(im1.commonbeam())
print(im2.commonbeam())
```
Gives two different beam sizes:
```
{'major': {'value': 4.211865471480522, 'unit': 'arcsec'}, 'pa': {'value': 29.99999999999998, 'unit': 'deg'}, 'minor': {'value': 3.658232394188034, 'unit': 'arcsec'}}
{'major': {'value': 4.343317398689631, 'unit': 'arcsec'}, 'pa': {'value': 29.25631919084969, 'unit': 'deg'}, 'minor': {'value': 3.5897767228849267, 'unit': 'arcsec'}}
```
Both of these aren't the optimal beam area. With the MVE approach I get a 4.49x3.28 arcsec beam with a PA of 30 deg.

Reply:
This is pretty counterintuitive.  If `common_2beams(beam1, beam2) != common_2beams(beam2, beam1)`, I'd argue we should *never* use `common_2beams`.

Reply:
Sorry, the order of a set of beams _larger than 2_ depends on the order. The comparison between two beams should always be fine.

Reply:
OK, so `common_2beams(common_2beams(beam1, beam2), beam3) != common_2beams(common_2beams(beam1, beam3), beam2)` is possible.